trinity/tools/async_process_runner.py

```diff
@@ -0,0 +1,64 @@
+import asyncio
+import os
+import signal
+from typing import (
+    AsyncIterable,
+    Awaitable,
+    Callable,
+    Iterable,
+)
+
+
+AwaitableGeneratorProducer = Callable[[], Awaitable[bytes]]
+
+
+class AsyncProcessRunner():
+
+    @classmethod
+    async def create_and_run(cls, cmds: Iterable[str], timeout_sec: int=10) -> 'AsyncProcessRunner':
+        runner = cls()
+        await runner.run(cmds, timeout_sec)
+        return runner
+
+    async def run(self, cmds: Iterable[str], timeout_sec: int=10) -> None:
+        proc = await asyncio.create_subprocess_exec(
+            *cmds,
+            stdout=asyncio.subprocess.PIPE,
+            stderr=asyncio.subprocess.PIPE,
+            # We need this because Trinity spawns multiple processes and we need to take down
+            # the entire group of processes.
+            preexec_fn=os.setsid
+        )
+        self.proc = proc
+        asyncio.ensure_future(self.kill_after_timeout(timeout_sec))
+
+    async def iterate_stdout(self) -> AsyncIterable[str]:
+        async for line in self._iterate_until_empty(self.proc.stdout.readline):
+            yield line
+
+    async def iterate_stderr(self) -> AsyncIterable[str]:
+        async for line in self._iterate_until_empty(self.proc.stderr.readline):
+            yield line
+
+    async def _iterate_until_empty(
+            self,
+            awaitable_bytes_fn: Callable[[], Awaitable[bytes]]) -> AsyncIterable[str]:
+
+        while True:
+            line = await awaitable_bytes_fn()
+            # Leaving this print here allows running pytest with -s and observing the output
+            # There's probably a better way to handle that
+            print(line)
+            if line == b'':
+                return
+            else:
+                yield line.decode('utf-8')
+
+    async def kill_after_timeout(self, timeout_sec: int) -> None:
+        await asyncio.sleep(timeout_sec)
+        self.kill()
+        raise TimeoutError('Killed process after {} seconds'.format(timeout_sec))
+
+    def kill(self) -> None:
+        # TODO: investigate if we could do this more gracefully
+        os.killpg(os.getpgid(self.proc.pid), signal.SIGKILL)
```

Comment:
look into using one of the utils from here: https://github.com/ethereum/py-evm/blob/master/trinity/utils/ipc.py#L18

Reply:
I removed the comment because I think we can actually keep it as is, because:

1. This is using `killpg` to kill the entire process group which isn't what our utils do
2. I *think* there is nothing we can gain from sending a `SIGTERM` first because these Trinity instances are throwaway onetime instances anyway. That means, the only effect a `SIGTERM` followed by a `SIGKILL` after some seconds would have are slower running tests.
3. Since I'm using `asyncio.subprocess.Process` I can't use `join` to wait on the shutdown and I believe waiting in an async fashion would complicate the tear down of the tests.

Overall, I think it's good as it is :)